tests/unit/test_artifacts.py

```
@@ -0,0 +1,216 @@
+    def test_nonexistent_root_fails_closed(self, tmp_path: Path) -> None:
+        with pytest.raises(ArtifactIntegrityError, match="root"):
+            verify_artifact(_manifest(), tmp_path / "does-not-exist")
+
+    def test_empty_manifest_rejected_at_construction(self) -> None:
+        with pytest.raises(ArtifactIntegrityError, match="at least one file"):
+            _manifest(entries=())
+
+    def test_directory_in_place_of_file_fails_closed(self, tmp_path: Path) -> None:
+        root = tmp_path / "dirroot"
+        (root / "alpha.txt").mkdir(parents=True)
+        with pytest.raises(ArtifactIntegrityError, match="regular file|missing"):
```

Ruff (0.16.2)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)